rust/kernel/allocator.rs

```diff
@@ -2,29 +2,136 @@
 
 //! Allocator support.
 
+use core::alloc::AllocError;
 use core::alloc::{GlobalAlloc, Layout};
 use core::ptr;
+use core::ptr::NonNull;
 
 use crate::bindings;
+use crate::static_assert;
 
-struct KernelAllocator;
+pub(crate) struct KernelAllocator;
+
+fn is_power_of_two(x: usize) -> bool {
+    (x != 0) && ((x & (x - 1)) == 0)
+}
+
+fn requires_header(layout: &Layout) -> bool {
+    layout.align() > bindings::ARCH_KMALLOC_MINALIGN as usize
+        && layout.size() > layout.align()
+        && !is_power_of_two(layout.size())
+}
+
+// Header containing a pointer to the start of an allocated block.
+// SAFETY: Size and alignment must be <= `bindings::ARCH_KALLOC_MINALIGN`.
+#[repr(C)]
+struct Header(*mut u8);
+
+static_assert!(core::mem::size_of::<Header>() <= bindings::ARCH_KMALLOC_MINALIGN);
+static_assert!(core::mem::align_of::<Header>() <= bindings::ARCH_KMALLOC_MINALIGN);
+
+impl KernelAllocator {
+    #[cfg(not(test))]
+    #[cfg(not(testlib))]
+    fn allocate_with_header(
```

Comment:
Perhaps a `#[cold]` attribute should be added to indicate this is an unlikely path.